### ISSUE TITLE
Reducing the feel of ink latency

### ISSUE THREAD

**Issue by 0ffz** [xournalpp/xournalpp#1902]
**Is your feature request related to a problem? Please describe.**
I've noticed the ink you write with trails a little behind the actual point of your cursor. This happens with my active pen and drawing tablet and I assume is caused by inevitable delay from rendering the ink. If I pause for a second, it eventually catches up, but it's very noticable when writing fast.

**Describe the solution you'd like**
Perhaps it would be possible to draw a temporary line from the last point where the ink is ending to the cursor for added responsiveness. 

Over on Windows, I swear OneNote goes a setup further and overshoots with this line so it's drawing where it thinks you will be. OneNote has always had this weird sort of snappy line while drawing, which I think is related to this. If that's the tradeoff they pay for responsiveness, I think it's worth it, and could be interesting to have as a configurable value here.

I haven't looked through any code here yet, but it sounds like a fun thing to play around with, though I'd like some insight into what the sources of latency are, and whether it's even worth attempting this sort of trick.

**Comment by Technius** [comment]
I didn't write the rendering code, but here's my understanding of how it works.
* Each page has its own buffer; all elements on the page are drawn on this page buffer. Any time an element is added, removed, or modified, the element will generate a request to be redrawn (more on this below).
* The process of drawing a page buffer to the actual canvas widget through GTK is called a "**repaint**."
* Each request to be redrawn is queued in the `Scheduler` as `RenderJob`s. A "**rerender**" will cause elements to be drawn to the page buffer and then queue a repaint of the page in GTK.
* Note that there is latency due to two stages of queueing: one delay for the render queue, and one delay for the repaint queue. 
* Typically, the element implementations use rerender more than repaint since changes need to drawn to the page buffer because they can be displayed on screen. Full page rerenders typically only occur on tool changes, zoom changes, or periodic sidebar preview updates.

I'm sure some of the other team members like @LittleHuba can provide additional insight on how the current rendering system works. We're planning on rewriting the rendering system later as we start to integrate the new file format, and hopefully it should be faster then. This may take a while though, so if you'd like to help find some sources of latency in the existing code, we'd really appreciate the help!

**Comment by LittleHuba** [comment]
@Technius did a great job highlighting the current setup. The new system will go a slightly different approach. 

Currently we only have one thread rendering changes. The new system will have at least four of these threads which should improve responsiveness. 

A quick breakdown of the workflow: 

- An element is modified which will notify the RenderingManager via a callback. The callback will enqueue the element in the rendering-queue.
- The element is picked up by the RenderingManager which renders the tiles (quadratic areas of a certain size) which cover the area the element is in. After rendering a tile it is stored for later access. 
- Each view has a DrawingManager that is notified by the RenderingManager when a tile is rerendered. If the tile is visible in the view the DrawingManager will redraw this tile. Since we have multiple threads we can afford doing this within the same thread that rendered this tile. This ensures cache-locality and saves us one queue. Since we are also buffering on our own, we will switch off buffered drawing in frameworks as much as we can. 

Some more details: 
- Full redraws of pages (scrolling, zooming,...) will use the already drawn tiles, which saves a lot of time and computation power. 
- Zooming requires rendering of correctly zoomed tiles. This will be done concurrently and slightly before they are actually required. This way we should be able to avoid any lag there. Additionally if tiles of a higher zoom-level are available we will do cheaper renders using a technique called Supersampling. 
- Since currently the system renders each single change it might be overwhelmed by the amount of points your hardware generates. The new system has a denouncing algorithm which does not enqueue a tile for rendering if it is already enqueued. 

Now to your question: 
This has certain drawbacks. Overshooting the drawn path would only be possible in the input system as the rendering would not be aware of anything like this. It does not even know which element was modified but only the tile that was. It can be even multiple elements that were modified in the same tile. 
We would generate at least one additional rendering-request per path and on cases where elements are drawn near the border of a tile force the system to draw two titles where actually only one is necessary. 
Overall this would mean more power consumption and more overall lag for rendering. I assume that would actually be what you don't want to acchieve. Also most letters have more curved lines than straight ones so overshooting would essentially be always off by a good margin. Long story short: We can try this out but I assume the experience would actually not be that great for the above reasons. 

Off-topic:
The snappiness of paths in OneNote is as far as I know from something else. OneNote does a poor prerender of the path while it is drawn. The path is an approximation of the actual one. I assume they are reducing vertices and such stuff to make it faster. When the path is finished it is rendered once more in a detailed way, which you see as "snapping". I actually hate this approach as it is really irritating when you are drawing fast as parts of the screen flash that were not touched by me at all...

**Comment by 0ffz** [comment]
I see, very interesting stuff, and I really appreciate the improvements you guys are working on! I guess in a way I was thinking of doing a poor render like OneNote does, but updating it live instead of just when the path is finished, though with how you've described the rendering system it does sound like it'd just result in more overall lag.

I think that pretty much answers any questions I had, so I guess feel free to close this issue. Best of luck on the rendering rewrite!

**Comment by daljit97** [comment]
Yeah OneNote uses input prediction to "guess" where the pen is going to move. I think xournalpp could use something like this based on the "velocity" of the current input and enable input prediction only when a certain threshold of velocity has been crossed. However, I think this is secondary and more focus should be given to improve the scroll and zoom performance. I haven't had a full look at the code, in particular it would be really nice that xournal would cache the input into some of sort of pixmap when a pinch to zoom event has started and only when the user release their fingers, repaint the entire page at a new scale. In this way, the only trade off would that the page would appear blurry when zooming but I think that is not too much of a problem.

**Comment by LittleHuba** [comment]
Actually the main issue is currently that we directly draw on the canvas of the view. So we can't make drawing and zooming concurrent. 
That is what the new system solves by introducing the tiles that are stored internally. We can render new titles with a higher zoom-level while the view still uses the old ones to update itself. That should go a long way in making zooming smooth. 
For scrolling we can render the adjacent tiles to the currently visible ones which will allow for smoother scrolling. Of course we have to find the sweet spot between increased computation and smoothness of scrolling there.

**Comment by gtusr** [comment]
Maybe incorporating this would help:

[Google's ink-stroke-modeler code](https://google.github.io/ink-stroke-modeler)

**Comment by Doublonmousse** [comment]
Well yes, some applications do a prediction for where the pen should be at each frame, bridging the gap between the last point recorded and the actual position of the pen (predicted).

The main use is not always latency reduction though. It's more that this allows to use smoothing methods without feeling too much additional lag (as most smoothing methods will trail behind and add latency for smoother lines).

Using the inertia preprocessor (see https://xournalpp.github.io/guide/config/preferences/) will already be somewhat close to what the google ink-stroke-modeler does (mass and spring system) without the prediction part
